SDGupdater

EmmaWoodONS opened the pull request:
merging SDGupdater into main for first time

EmmaWoodONS commented:
Hi @ANikolova22 - do you have time to look at this? I split it off from the 3-2-2_and_SDGupdater branch, as it wasn't really right they were on the same branch. I have addressed the relevant comments from the review of that branch, and I am pretty sure all the functions work, and there are at least basic unit tests for everything. It would be good to check that it installs correctly, as I am not 100% sure the build worked correctly (if it can see all of the functions I think all is well). If you are tight on time I will ask Varun or David to pick it up.

ANikolova22 commented:
Hi @EmmaWoodONS, yes I'll aim to have it done by end of this week of that's ok?

EmmaWoodONS commented:
> Hi @EmmaWoodONS, yes I'll aim to have it done by end of this week of that's ok?

That would be awesome - thanks Atanaska!

ANikolova22 commented:
Sorry for the massive delay on this! I probably could have done a more throughout job, but this is a start... I'll add in-line comments for the more specific bits, but below is an overall summary

- Generally, the package installed without a problem. I had to navigate to the tests folder to be able to run the bulk test test_check("SDGupdater"), perhaps because I don't have it set up as an R project; Only one test fails: `test_remove_blanks_and_info_cells`, and there are 14 warnings.
- `remove_blanks_and_info_cells` returns an error for test "_if input does not look like it comes from xlsx_cells input_" but not sure why - I tested it by trying to feed a .csv file and I got the correct error message for the function. Maybe I misunderstood and that error IS expected because wrong input is given on purpose in the test file, in which case it's all good...
- Some of the data downloadable from ONS page is **xls**, not **xlsx** so may need to make a note that the user needs to 'save as' for the latest format (otherwise the `xlsx_cell()` function won't work)
- 14 warnings in total, found in all the superscript-related test files, also test for `get_all_years returns expected value`, and `get_characters_after_dot returns expected data datatype`  - not sure if it's an issue or this is expected, the warnings are related to wrong data types
- `collapse_list_to_strings() `- the example in the function docs uses the old name of the function (list_into_vector_of_strings)
- `remove_superscripts() `may need more fleshing out in the documentation for the different usages, e.g. what is meant by false superscripts? Also, the value returned may differ based on which variation of the function is used - `could_contain_superscript()` returns a vector of logicals, rather than a character vector (as currently specified in the documentation). 
- Unlikely scenario, but if a character starts with a number or a symbol and has a space, the possible superscript does not get detected. For example  `remove_superscripts("2 inhabitants1") `or `remove_superscripts("% inhabitants1")` don't do anything, but `remove_superscripts("Two inhabitants1")` gives "Two inhabitants". If the digit/symbol is in the middle of the character (with spaces) seems to be working fine (e.g. `remove_superscripts("two % inhabitants1")`)
- `count_mismatches()` doesn't work as expected because the ifelse is reading the wrong column name in the function code (easy fix, see in-line comment)
- `multiple_country_warning` in `get_all_country_names()` doesn't seem to work, but I can't quite work out what exactly it should be doing
- `get_fourth_from_last_character() `description says the argument must be a character string, but it actually works with numeric variables too (integers and doubles)
- The example syntax for `get_modal_values() ` is not formatted correctly - the test numbers are not fed as a vector so it comes up with an error. Should be `get_modal_values(c(1,1,2,3)) `not `get_modal_values(1,1,2,3)`. All the tests pass though. But what about data without a mode? `get_modal_values(c(1,2,3)) `returns "1,2,3", but maybe it shouldn't?
- `get_penultimate_character()` is not limited to strings, as the description states. Numeric variables also work
- `not_in()` seems incomplete?
- `unique_to_string() `works with numeric vectors too, not just strings

ANikolova22 commented on SDGupdater/R/count_mismatches.R:
```diff
@@ -0,0 +1,27 @@
+
+  data_frame <- data.frame(var1 = number_var_1, var2 = number_var_2)
+  no_NAs <- na.omit(data_frame)
+  comparison <- ifelse(no_NAs$number_var_1 == no_NAs$number_var_2, 0, 1)
```
Line 23 should be `comparison <- ifelse(no_NAs$var1 == no_NAs$var2, 0, 1)`

ANikolova22 commented on SDGupdater/R/get_modal_values.R:
```diff
@@ -0,0 +1,23 @@
+#' @return Vector of same type as values.
+#'
+#' @examples
+#' get_modal_values(1,2,3,1,3)
```
Lines 14, 15, and 16 for the examples, the numbers should be fed as a vector, so `get_modal_values(c(1,2,3,1,3))`

EmmaWoodONS commented:
> * `remove_blanks_and_info_cells` returns an error for test "_if input does not look like it comes from xlsx_cells input_" but not sure why - I tested it by trying to feed a .csv file and I got the correct error message for the function. Maybe I misunderstood and that error IS expected because wrong input is given on purpose in the test file, in which case it's all good...

This was failing because I did expect_warning rather than expect_error. It now passes.

EmmaWoodONS commented:
> * Some of the data downloadable from ONS page is **xls**, not **xlsx** so may need to make a note that the user needs to 'save as' for the latest format (otherwise the `xlsx_cell()` function won't work)

Yep - good point. I have made sure to put this in the readme files so far, but have also added it to that error message.

EmmaWoodONS commented:
> * 14 warnings in total, found in all the superscript-related test files, also test for `get_all_years returns expected value`, and `get_characters_after_dot returns expected data datatype`  - not sure if it's an issue or this is expected, the warnings are related to wrong data types

warnings have all been fixed now - they were all correct, but I have added `suppressWarnings(function)` so that the tests pass

EmmaWoodONS commented:
> collapse_list_to_strings() - the example in the function docs uses the old name of the function (list_into_vector_of_strings)

yep - good spot - thanks! I have now written basic unit tests for this one and fixed the issue

EmmaWoodONS commented:
>`remove_superscripts()` may need more fleshing out in the documentation for the different usages, e.g. what is meant by false superscripts? Also, the value returned may differ based on which variation of the function is used - `could_contain_superscript()` returns a vector of logicals, rather than a character vector (as currently specified in the documentation).

Thanks - have now fleshed out the documentation

EmmaWoodONS commented:
> `count_mismatches() `doesn't work as expected because the ifelse is reading the wrong column name in the function code (easy fix, see in-line comment)

yep - good spot - thanks! I have now written basic unit tests for this one and fixed the issue

EmmaWoodONS commented:
> `multiple_country_warning` in `get_all_country_names()` doesn't seem to work, but I can't quite work out what exactly it should be doing

Thanks  - I missed some important stuff in the documentation and example (you need to specify a country string). I have also now written a basic unit test for this function so it should be clearer. I may well have been overdoing it with including this as a function, but I was tired of writing out the same warning message!

EmmaWoodONS commented:
> `get_fourth_from_last_character()` description says the argument must be a character string, but it actually works with numeric variables too (integers and doubles)

Good spot. Now fixed in documentation and extra tests added

EmmaWoodONS commented:
> The example syntax for get_modal_values() is not formatted correctly - the test numbers are not fed as a vector so it comes up with an error. Should be get_modal_values(c(1,1,2,3)) not get_modal_values(1,1,2,3). All the tests pass though. But what about data without a mode? get_modal_values(c(1,2,3)) returns "1,2,3", but maybe it shouldn't?

Thanks :) Now corrected in example. The mode of 1, 2, 3 (and 1, 1, 2, 2, 3, 3 etc) is 1, 2, 3, so that is fine. I have added a not in the documentation that it works for multi-modal data, and returns all modes.

EmmaWoodONS commented:
> `not_in()` seems incomplete?

I didn't have much to say about it - maybe there is a better way to document operators? I was just tired of writing this at the top of every script!
